### ISSUE TITLE
Triton kernel unexpectedly gets 1.35x slower by more specializaiton

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#120667]
This triton kernel https://gist.github.com/shunting314/9bcd806bdcda7f58ab4bae57ab6f80ad (call it v1) uses ks0 as stride. If we replace ks0 with the only possible value in this test, the new kernel https://gist.github.com/shunting314/063eb3c6bf16c1b947d7608eae87381c (call it v2) get 1.35x slower.

I can verify that the slower kernel uses vectorized load/store while the faster does not. This is very counter-intuitive since vectorized load/store should help perf in general.

Actually there is an alternative change to make v2 kernel as faster as v1: we can remove in_ptr0 and out_ptr1 from the divisible_by_16 list. This by effect disables vectorized load/store as well. Here is the kernel (v3): https://gist.github.com/shunting314/827b482a0407425565051538633aa095

Overall the numbers I get seems to suggest that vectorized load/store here is bad for perf.

@bertmaher  @htyu  let me know if I should cut the issue in the triton repo instead.

UPDATE TO AVOID CONFUSION: the root cause of the slowness is not vectorization. Vectorization still help in general. The root cause is the slow kernel uses an sub-optimal layout that harm memory coalescing. 

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @jansel @Chillee @eellison

**Comment by jansel** [comment]
Weird... if we don't find a fix maybe we should open an issue to openai/triton also.

**Comment by htyu** [comment]
Yeah, it's counter-intuitive. I'm taking a look at what's exactly going on.

**Comment by htyu** [comment]
> I can verify that the slower kernel uses vectorized load/store while the faster does not. 

I'm seeing this too. In addition, I'm seeing `fma` instruction is used for the first loop in the fast version while it's not in the slower version. I guess that causes the perf gap without confirming it in the profiler. This is likely an issue with the NV LLVM backend.

So it seems that without vectorization, the memory accesses from different threads can still be coalesced, resulting the same number of memory transactions. While the vectorized version brings more work to a single thread, but except for the loads, the rest of operations aren't very effective though.

**Comment by htyu** [comment]
> What's the CUDA grid size in case of vectorized / non-vectorized variants?

Same, i.e, 8192 below

      `  kernel_info = triton_.run(*args, 8192, 50272, grid=grid(8192), stream=stream0)`

Same number of threads are used for both cases.

**Comment by shunting314** [comment]
yea, both kernels (v1 and v2) uses XBLOCK==1, so the grid size is (8192, 1, 1).  num_warps is 8 for both kernels so the block dims are (256, 1, 1). This should match what you see in ncu
```
  triton__0d1d2d3d4d5d6d78de9 (8192, 1, 1)x(256, 1, 1), Context 1, Stream 7, Device 0, CC 8.0
```

**Comment by shunting314** [comment]
In case it helps, here is a script https://gist.github.com/shunting314/74f5d1b4cd9c66ba5ad7cb3896fd9c1c to run the fast and slow kernel side by side . cc @htyu

**Comment by htyu** [comment]
> In case it helps, here is a script https://gist.github.com/shunting314/74f5d1b4cd9c66ba5ad7cb3896fd9c1c to run the fast and slow kernel side by side . cc @htyu
> 
> Update: here is a slightly different version but much simpler https://gist.github.com/shunting314/142924455be75bd393697499d3c0d09e . I removed the unnecessary parts that does not affect repro. Now the vectorized version took 6.032 ms while the non-vectorized version takes 4.586 ms

Thanks. I'll give it a shot. 

BTW, is there a way to map the triton kernels back to the python source?

**Comment by htyu** [comment]
> In case it helps, here is a script https://gist.github.com/shunting314/74f5d1b4cd9c66ba5ad7cb3896fd9c1c to run the fast and slow kernel side by side .

I gave this a shot and in the profiler I did see the vectorized kernel had lower compute throughput with same memory throughput.

**Comment by shunting314** [comment]
> BTW, is there a way to map the triton kernels back to the python source?

There is no automatic way so far. The kernel may not map to a python function directly since they may contains part of decompostion of some ops and also fuses those fragments from different ops. Or they may contains some backward part depends on how to partitioner works.

But if you want, you can manually do the conversion. The simplified kernel I pasted just do some simple computation from tensor a, b, c and write the result to d. Should be trivial to map them back to pytorch kernel if we want.

**Comment by htyu** [comment]
An update after discussing with Triton upstream in https://triton-lang.slack.com/archives/C06EJRSF7CN/p1709668178748089 . This sounds a suboptimal heuristic in Triton about memory coalescing. I'm going to dig more.

**Comment by jansel** [comment]
@BoyuanFeng it doesn't look like that PR is landed?  

@htyu is this still an issue?

**Comment by htyu** [comment]
Right, the PR isn't landed. It should still be an issue.

The reason that PR wasn't landed because the heuristics there didn't work across the board, and the benchmark it regressed was from OpenAI and was not available to us.

**Comment by masnesral** [comment]
[old issue scrubbing] @htyu is this still on your radar? I tried the repros myself, but it looks like they're a bit out of date with the helpers in runtime/triton_heuristics and I'm not sure about the original code we used to generate these kernels.
